File: src/deepsearch/search/web.py
from dataclasses import dataclass

from duckduckgo_search import DDGS


@dataclass(frozen=True)
class SearchHit:
    title: str
    url: str
    snippet: str


def search_web(query: str, max_results: int) -> list[SearchHit]:
    with DDGS() as ddgs:
        raw = list(ddgs.text(query, max_results=max_results))

    return [
        SearchHit(
            title=item.get("title", ""),
            url=item.get("href", ""),
            snippet=item.get("body", ""),
        )
        for item in raw
        if item.get("href")
    ]


def format_hits(hits: list[SearchHit]) -> str:
    if not hits:
        return "(no results)"

    blocks: list[str] = []
    for index, hit in enumerate(hits, start=1):
        blocks.append(
            f"[{index}] {hit.title}\nURL: {hit.url}\n{hit.snippet}".strip()
        )
    return "\n\n".join(blocks)

File: src/server/search/web.ts
import { search } from "duck-duck-scrape";

export interface SearchHit {
  title: string;
  url: string;
  snippet: string;
}

export async function searchWeb(
  query: string,
  maxResults: number,
): Promise<SearchHit[]> {
  const response = await search(query);
  return response.results.slice(0, maxResults).flatMap((item) => {
    if (!item.url) return [];
    return [
      {
        title: item.title ?? "",
        url: item.url,
        snippet: item.description ?? item.rawDescription ?? "",
      },
    ];
  });
}

export function formatHits(hits: SearchHit[]): string {
  if (hits.length === 0) return "(no results)";

  return hits
    .map(
      (hit, index) =>
        `[${index + 1}] ${hit.title}\nURL: ${hit.url}\n${hit.snippet}`.trim(),
    )
    .join("\n\n");
}
